Implement composable UI for Dynamic Dashboard Cards

justtwago opened the pull request:
Fixes #19735

This PR includes composable UI for the dynamic dashboard cards feature only. Unfortunately, there is no way to load images in the preview mode, so I created placeholders for icons and feature images. To test with  We should consider if we are good to have such placeholders or modify them.

<img width="650" alt="Screenshot 2023-12-12 at 3 06 34 PM" src="https://github.com/wordpress-mobile/WordPress-Android/assets/16563318/ed1cb35d-06ca-44f2-a1c6-ebbf6573e99b">

-----

## To Test:

Go to `org/wordpress/android/ui/mysite/cards/dynamiccard/DynamicCardComposable.kt` file in AndroidStudio and make sure the cards preview look accordingly to the designs: YSMw2nbFAgPpjZLLbYvTfT-fi-1281_3822

-----

## Regression Notes

1. Potential unintended areas of impact

    - N/A

2. What I did to test those areas of impact (or what existing automated tests I relied on)

    - N/A

3. What automated tests I added (or what prevented me from doing so)

    - N/A

-----

## PR Submission Checklist:

- [ ] I have completed the Regression Notes.
- [ ] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

## UI Changes Testing Checklist:

- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

Commit:
Implement composable UI for Dynamic dashboard cards feature

Commit:
Refactor DynamicCardComposable to split composables

antonis commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteCardAndItem.kt:
```diff
@@ -389,8 +391,27 @@ sealed class MySiteCardAndItem(open val type: Type, open val activeQuickStartIte
             val onCtaClick: ListItemInteraction,
         ) : Card(type = Type.WP_SOTW_2023_NUDGE_CARD)
 
-        data class NoCardsMessage(val title: UiString, val message: UiString)  : Card(Type.NO_CARDS_MESSAGE)
```
Thank you for fixing those indentation issues 🙇

antonis commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteCardAndItem.kt:
```diff
@@ -389,8 +391,27 @@ sealed class MySiteCardAndItem(open val type: Type, open val activeQuickStartIte
+            val image: String?,
+            val action: String?,
+            val onHideMenuItemClick: ListItemInteraction,
+            val onCtaClick: ListItemInteraction,
```
This is not a suggestion for a change, I'm just thinking out loud regarding the nullability of the `onCtaClick`.
I was thinking that:
* The `onCtaClick` should be non-null when we have an `action` string and a `url` from the model https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2923/files#diff-0834739cb5282253347464dba9388f8adc41facaa790a9fc28b0e5e64511ce1cR72
* When one of those is null there is no action and the action button should be hidden even if we have a string for it.

Wdyt?

justtwago replied:
Totally makes sense. I revamped the feature according to this suggestion: p1702484824130069/1702473203.210019-slack-C067H36F011

antonis reviewed (commented):
Awesome work @justtwago 👍 
The code looks great and the UI aligns with the design 🎉 
I noticed just an issue with the `Dark Mode` when testing on a device.

|Light Mode|Dark Mode|
|---|---|
|![Screenshot_20231213_135819](https://github.com/wordpress-mobile/WordPress-Android/assets/304044/3743993f-1959-46ff-b53f-96a981a7d087)|![Screenshot_20231213_140001](https://github.com/wordpress-mobile/WordPress-Android/assets/304044/08b1219b-17bf-4d32-99f3-00e3fefa3a7d)|

antonis commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/dynamiccard/DynamicCardCallToActionButton.kt:
```diff
@@ -0,0 +1,28 @@
+    ) {
+        Text(
+            text = text,
+            style = MaterialTheme.typography.bodyMedium.copy(
```
Would it make sense to reuse [the `DashboardCardTypography`](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/WordPress/src/main/java/org/wordpress/android/ui/compose/styles/DashboardCardTypography.kt) like other dashboard cards (e.g. [the recent `BloganuaryNudgeCard`](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/dashboard/bloganuary/BloganuaryNudgeCard.kt)) or that would require falling back to `AppTheme` instead of the `M3Theme` 🤔

justtwago replied:
All in all, I migrated the entire feature to AppTheme and used [DashboardCardTypography](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/WordPress/src/main/java/org/wordpress/android/ui/compose/styles/DashboardCardTypography.kt) to not mix up so many different approaches: Android View, M2 Compose, and M3 Compose.

antonis replied:
Makes sense 👍 
Thank you for the change Artyom 🙇

Commit:
Migrate dynamic cards back to Material 2 to use it them with AppTheme

Commit:
Revamp CTA logic:
- Introduced CTA on the entire card when an action title is null but an URL is not
- Introduce nullability of a CTA

Commit:
Remove a redundant order property from the dynamic card UI model

justtwago commented:
Thank you for reviewing the PR, @antonis!
I implemented the suggestions and fixed the issue with a dark mode.

justtwago commented on WordPress/src/main/java/org/wordpress/android/ui/compose/components/card/UnelevatedCard.kt:
```diff
@@ -6,21 +6,23 @@ import androidx.compose.material.Card
+    val shape = RoundedCornerShape(10.dp)
     Card(
-        modifier = modifier,
+        modifier = Modifier.clip(shape).then(modifier),
```
This adds rounded corners to a click ripple effect.

antonis replied:
TIL :)

justtwago commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/MySiteCardAndItem.kt:
```diff
@@ -389,8 +391,39 @@ sealed class MySiteCardAndItem(open val type: Type, open val activeQuickStartIte
+                val description: String?,
+            )
+
+            sealed class ActionSource {
+                abstract val url: String
+                abstract val onCtaClick: ListItemInteraction
+
+                data class Card(
+                    override val url: String,
+                    override val onCtaClick: ListItemInteraction
+                ) : ActionSource()
+
+                data class Button(
+                    override val url: String,
+                    override val onCtaClick: ListItemInteraction,
+                    val title: String
+                ) : ActionSource()
+            }
```
According to p1702484824130069/1702473203.210019-slack-C067H36F011:
- URL is mandatory for a CTA
- CTA button is visible when there is a title
- Card CTA is available for actions with no title
- onCtaClick property available for non-null action only

antonis replied:
Thank you for implementing this Artyom 🙇

antonis commented on WordPress/src/main/java/org/wordpress/android/ui/mysite/cards/dynamiccard/DynamicCardComposable.kt:
```diff
@@ -118,7 +169,27 @@ fun DynamicDashboardCardWithFeatureAndSubtitleAndDescriptionPreview() {
+
+@Preview
+@Composable
+fun DynamicDashboardWithFeatureImageOnly() {
```
Thank you for adding extra previews 🙇

antonis reviewed (approved):
Awesome work @justtwago ⭐ 
Thank you for implementing the CTA spec changes and fixing dark mode. I retested the app and everything worked as expected for me. The code also looks great 🎉
